R/efficacy_nat.R

```diff
@@ -1,129 +1,142 @@
-    ef_transmission[finite_ab] <- 1 - ef_transmission[finite_ab]
-    return(ef_transmission)
-  } else {
-    return(ef_transmission)
-  }
+  ef_transmission <- rep(1, length(nat))
+
+  nonzero_nat <- which(nat > 0)
+
+  ef_transmission[nonzero_nat] <- 1 / (1 + exp(-parameters$k * (log10(nat[nonzero_nat] / parameters$nt_transmission_factor) - log10(parameters$ab_50)))) # reported efficacy in trials
+  ef_transmission[nonzero_nat] <- 1 - ef_transmission[nonzero_nat]
+
+  return(ef_transmission)
 }
 
 
 #' @title Make function to calculate population NAT
-#' @description This returns a function taking two arguments `variables` (a list of variables)
-#' and `index` (a [individual::Bitset]) which will return the NAT for each person in `index`.
-#' If the list of variables includes `ab_titre_inf`, then the combined NAT from vaccine and
-#' infection derived NATs is returned.
+#' @description The returned function takes arguments `index` (a bitset) and `day` and returns the NAT of persons in `index` on the linear scale.
 #' @param variables a named list
+#' @param parameters a named list
 #' @export
-make_calculate_nat <- function(variables) {
+make_calculate_nat <- function(variables, parameters) {
   stopifnot(!is.null(variables$ab_titre))
   stopifnot(inherits(variables$ab_titre, "DoubleVariable"))
   if (is.null(variables$ab_titre_inf)) {
 
-    calculate_nat <- function(variables, index) {
-      variables$ab_titre$get_values(index)
-    }
+    stop("currently safir only correctly simulates models with seperate tracking of vaccine and infection derived NAT values")
 
   } else {
     stopifnot(inherits(variables$ab_titre_inf, "DoubleVariable"))
 
-    calculate_nat <- function(variables, index) {
-      # two types of NAT
-      nat_vaccine <- variables$ab_titre$get_values(index)
-      nat_infection <- variables$ab_titre_inf$get_values(index)
-      # add them for single effect
-      nat <- exp(nat_vaccine) + exp(nat_infection)
-      nat <- log(nat)
-      return(nat)
+    # for new variant proof VFR modeling
+    if (!is.null(parameters$vp_time)) {
+
+      dt <- parameters$dt
+
+      calculate_nat <- function(index, day) {
+
+        # Get VFR parameter, if null assign 1
+        if (is.null(parameters$vfr)) {
+          vfr <- 1
+        } else {
+          vfr <- parameters$vfr[day]
+        }
+
+        nat_vaccine <- exp(variables$ab_titre$get_values(index))
+        nat_infection <- exp(variables$ab_titre_inf$get_values(index))
+
+        # nat_infection should always be scaled by vfr regardless
+        nat_infection <-  pmax(.Machine$double.eps, nat_infection / vfr)
+
+        # find individuals who were vaccinated when variant proof vaccine was on
+        dose_times <- ceiling(variables$dose_time$get_values(index) * dt)
+
+        been_vaccinated <- variables$dose_num$get_index_of(set = 0)$not()
```

Comment:
This line I think may be not quite right. 

`index` for this function is all those who are susceptible and is thus not necessarily all individuals covered by variables. So this current line is getting out of everyone who has been vaccinated.  Should it not be getting who has been vaccinated amongst `index`. 

Would something like this be correct? (this would replace line 101 and 103_
```suggestion
        been_vaccinated_vec <- which(variables$dose_num$get_values(index) != 0)
```